---
title: "fitting_dPCR_data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{fitting_dPCR_data}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

In this document, we show how the likelihood model for duplex dPCR assays derived in Caduff et al. 2021 [(doi.org/10.1101/2021.08.22.21262024)](doi.org/10.1101/2021.08.22.21262024) can be used to compute confidence intervals for individual measurements, as well as fitting logistic growth curves and inferring growth rates for time series data. The logistic growth here starts with a >0 background prevalence. 


```{r setup}
library(WWdPCR)
```

# Example on simulated data

## Simulate dPCR data

Will simulate 10 ddPCR measures of a drop-off duplex assay. We assume a fixed 10,000 droplets per assay. We vary randomly the (low) concentration of total RNA. We then draw counts of: double negative droplets, single positive droplets, double positive droplets. 

```{r}
set.seed(42)

# parameters

TT <- 10 #timesteps
a <- 0.5 #parameters of 3PL
b <- 8
c <- 0.1
N <- 10000 #number of cells

# true curve of the ratios
rs <- pl3model(1:TT, a,b,c) 

# draw random concentrations
lambdas <- abs(rnorm(TT, 0.02, 0.01)) # concentrations

# draw droplet counts 
counts <- lapply(1:TT, function(t){
    rmultinom(1, N, c(exp(-lambdas[t]),
                      1-exp(-(rs[t])*lambdas[t]),
                      exp(-(rs[t])*lambdas[t])-exp(-lambdas[t])))
})
counts <- t(Reduce(function(...) cbind(...), counts))
```


## Fit logistic growth models

Fit the (appropriate) three parametric logistic growth model (3PL), with a $c$ lower asymptote to account for the background prevalence, i.e. the ratio of the mutation follows:

$$
r_{m u t}(t)=c+(1-c) \frac{e^{a\left(t-t_{0}\right)}}{1+e^{a\left(t-t_{0}\right)}}
$$

```{r}
fit_3PL <- fit_pl(counts, # ddPCR droplet counts
                  1:TT, # timepoints
                  likelihood = loglik_trinom_prof, # likelihood for duplex dPCR data
                  fitfun = pl3model, # 3PL logistic growth model
                  quasi=FALSE, 
                  starting_values = c(a = 0.05, b = 10, c = 0.2))
knitr::kable(confint(fit_3PL))
```

Fit the (inappropriate) two parametric logistic growth model (2PL), without taking into account the background prevalence, i.e. the ratio of the mutation follows:


$$
r_{m u t}(t)=\frac{e^{a\left(t-t_{0}\right)}}{1+e^{a\left(t-t_{0}\right)}}
$$

```{r}
fit_2PL <- fit_pl(counts, # ddPCR droplet counts
                  1:TT, # timepoints
                  likelihood = loglik_trinom_prof, # likelihood for duplex dPCR data
                  fitfun = pl2model, # 2PL logistic growth model
                  quasi=FALSE, 
                  starting_values = c(a = 0.05, b = 10))
knitr::kable(confint(fit_2PL))
```

We see that the growth rate $a$ is underestimated when not taking into account the background prevalence. The fact that the $c$ parameter in the 3PL model has a Wald 95% confidence interval that does not include zero already shows us that this parameter should be included, but we can more formally compare the two models with a likelihood-ratio test:

```{r}
# compute quasi log-likelihood ratio, with inflation factor from 2PL (more conservative)
llr <- -2*((-1*fit_3PL$likelihood) - (-1*fit_2PL$likelihood))
# chisquare test of the log-likelihood ratio
print(c("2logLr:"=round(llr,2), "p-val:"=formatC(1-pchisq(llr, 1), format = "e", digits = 2)) )

```


## Plot

We will plot only the correct 3PL fit, along with the true curve. We will also plot 95% likelihood-ratio based confidence intervals for the individual data points. 

```{r, fig.dim=c(7, 5)}
# compute and plot empircal frequencies in data
empirical_freqs <- r_hat(counts)

# plot
plot(1:TT, empirical_freqs, ylim=c(0,1), main = "simulated data", ylab = "freq. mutation", xlab = "time")

# compute and plot likelihood ratio based confints for empirical freqs
empirical_freqs_confint <- lrt_confint(counts)
arrows(1:TT, 
       empirical_freqs_confint$lower, 
       1:TT, 
       empirical_freqs_confint$upper,
       length=0.05, angle=90, code=3)

# plot true curve and fitted curves
lines(1:TT, rs, col="black")
lines(1:TT, fit_3PL$fitted, col="red")


# make confidence bands through logit reparametrization and plot them
predicted_3PL_logit <- predict(fit_3PL, scale="logit", se=TRUE)
polygon(c(1:TT, rev(1:TT)),
        c(logit_inv(predicted_3PL_logit$pred - 1.96*predicted_3PL_logit$se),
          rev(logit_inv(predicted_3PL_logit$pred + 1.96*predicted_3PL_logit$se))),
       col=rgb(red=1, green=0, blue=0, alpha=0.1), border=NA)

# make legend
legend("topleft", legend=c("true", "3PL fit", "empirical"),
       col=c("black", "red", "black"), lty=c(1,1, NA), pch = c(NA, NA, 1), cex=.8)

```


# Fit ZH dPCR data

## Load data

We will use wastewater RT-ddPCR data from Caduff et. al. [(doi.org/10.1101/2021.08.22.21262024)](doi.org/10.1101/2021.08.22.21262024). The assay targets SARS-CoV-2 S-gene HV69-70 deletion, a signature of the B.1.1.7 (Alpha) variant. However, the deletion is also present in other lineages, and thus has a background prevalence. In the attached dataset, x0 x1 and x2 are the counts of double negative, single positive and double positive droplets, respectively. 

```{r}
str(dPCRdat_HV6970)
```

## Fit logistic growth models

Fit 3PL:

```{r}
dPCR_HV6970del_3PL <- fit_pl(dPCRdat_HV6970[,c("x0", "x1", "x2")],
                             dPCRdat_HV6970$dayssince,
                             likelihood = loglik_trinom_prof,
                             fitfun = pl3model,
                             quasi=TRUE,
                             starting_values = c("a"=0.05, "b"=50, "c"=0.01),
                             lower=c(0, -Inf, 0), upper=c(1, Inf, 1))
knitr::kable(confint(dPCR_HV6970del_3PL))
```

Fit 2PL:

```{r}
dPCR_HV6970del_2PL <- fit_pl(dPCRdat_HV6970[,c("x0", "x1", "x2")],
                             dPCRdat_HV6970$dayssince,
                             likelihood = loglik_trinom_prof,
                             fitfun = pl2model,
                             quasi=TRUE,
                             starting_values = c("a"=0.05, "b"=50),
                             lower=c(0, -Inf, 0), upper=c(1, Inf, 1))
knitr::kable(confint(dPCR_HV6970del_2PL))
```

We can compare it to a fit on the number of clinical B.1.1.7 cases. The data from Swiss clinical sequencing is attached, and was obtained from [Cov-Spectrum](https://cov-spectrum.ethz.ch/).

```{r}
str(clinical_data)
# let's aggregate them by date for all "divisions" (Cantons)
clinical_data_tot <- aggregate(. ~ date, data=clinical_data, sum)
clinical_data_tot <- clinical_data_tot[(as.Date(clinical_data_tot$date) >= as.Date("2020-12-07")) & 
                                           (as.Date(clinical_data_tot$date) <= as.Date("2021-03-26")), ]
clinical_data_tot$days_numeric <- as.numeric(as.Date(clinical_data_tot$date)-as.Date("2020-12-07"))

clinical_B117_2PL <- fit_pl(clinical_data_tot$B117 / clinical_data_tot$Tot, 
                               clinical_data_tot$days_numeric, 
                               clinical_data_tot$Tot,
                               fitfun = pl2model,
                               quasi=T, 
                               starting_values = c("a"=0.1, "b"=50), 
                               lower=c(0, -Inf, 0), upper=c(1, Inf, 1))
knitr::kable(confint(clinical_B117_2PL))
```

The growth rate estimate based on the 3PL fit on the wastewater dPCR data is closer to the growth rate estimate based on Swiss clinical sequencing data, than the growth rate estimate based on the 2PL fit on the wastewater dPCR data, which seems to suffer from underestimation due to background prevalence of the S:HV69-70 deletion. 


## Plot

```{r, fig.dim=c(7, 5)}
# compute empircal frequencies in data
dPCR_HV6970del_empirical_freqs <- r_hat(dPCRdat_HV6970[,c("x0", "x1", "x2")])

# plot
plot(dPCRdat_HV6970$date, dPCR_HV6970del_empirical_freqs, ylim=c(0,1),
     xlab = "date", ylab = "frac. S:HV69-70 deletion", 
     main="ZÃ¼rich Wastewater Samples")
lines(dPCRdat_HV6970$date, dPCR_HV6970del_3PL$fitted, col="red")

# compute and plot likelihood ratio based confints for empirical freqs
empirical_freqs_confint <- lrt_confint(dPCRdat_HV6970[,c("x0", "x1", "x2")])
arrows(dPCRdat_HV6970$date, 
       empirical_freqs_confint$lower, 
       dPCRdat_HV6970$date, 
       empirical_freqs_confint$upper,
       length=0.05, angle=90, code=3)

# make confints through logit reparametrization and plot them
predicted_dPCR_HV6970del_3PL_logit <-  predict(dPCR_HV6970del_3PL, scale="logit", se=TRUE)
polygon(c(dPCRdat_HV6970$date, rev(dPCRdat_HV6970$date)),
        c(logit_inv(predicted_dPCR_HV6970del_3PL_logit$pred - 1.96*predicted_dPCR_HV6970del_3PL_logit$se),
          rev(logit_inv(predicted_dPCR_HV6970del_3PL_logit$pred + 1.96*predicted_dPCR_HV6970del_3PL_logit$se))),
       col=rgb(red=1, green=0, blue=0, alpha=0.1), border=NA)

```


## Inspect overdispersion

We can inspect the overdispersion. It seems constant w.r.t sample size: the quasimultinomial model seems to be adequate in that respect. 

```{r, fig.dim=c(7, 5)}
n <- apply(dPCRdat_HV6970[,c("x0", "x1", "x2")], 1, sum)
plot(n, log10(dPCR_HV6970del_3PL$overdispersion[[1]]),
    xlab="sample size",
    ylab="log10 variance ratio",
    main="overdispersion vs sample size",
    ylim=range(log10(unlist(dPCR_HV6970del_3PL$overdispersion))))
points(n, log10(dPCR_HV6970del_3PL$overdispersion[[2]]),
       col="red")


for (i in 1:100){
    indxs <- sample(1:length(dPCR_HV6970del_3PL$overdispersion[[1]]), replace=T, prob=n)
    lines(loess.smooth(n[order(n)][indxs], log10(dPCR_HV6970del_3PL$overdispersion[[1]])[order(n)][indxs]), col=rgb(red=0, green=0, blue=0, alpha=0.2))
}
for (i in 1:100){
    indxs <- sample(1:length(dPCR_HV6970del_3PL$overdispersion[[1]]), replace=T)
    lines(loess.smooth(n[order(n)][indxs], log10(dPCR_HV6970del_3PL$overdispersion[[2]])[order(n)][indxs]),
          col=rgb(red=1, green=0, blue=0, alpha=0.2))
}
lines(loess.smooth(n[order(n)], log10(dPCR_HV6970del_3PL$overdispersion[[1]])[order(n)]), col="black", lwd=2)
lines(loess.smooth(n[order(n)], log10(dPCR_HV6970del_3PL$overdispersion[[2]])[order(n)]), col="red", lwd=2)


points(n, log10(dPCR_HV6970del_3PL$overdispersion[[1]]),
       col="black")
points(n, log10(dPCR_HV6970del_3PL$overdispersion[[2]]),
       col="red")

legend("bottomleft", legend=c(expression(x[1]), expression(x[2])),
       col=c("black", "red"), lty=c(1,1), pch=c(1,1), cex=1)
```


